### ISSUE TITLE
Phinx should safely handle concurrent migrations

### ISSUE THREAD

**Issue by markkimsal** [cakephp/phinx#1363]
Moving a deployment pipeline to load balanced machines has introduced the problem of not being able to safely run "phinx migrate" from all web servers.  This needlessly complicates the push pipeline because now a single machine has to be tagged as the one machine to perform DB migrations.

If phinx obtained a write lock on `phinxlog` (before `getVersionLog()` is called), only one connection would obtain the lock, causing all other connections to queue and wait for the migration to finish.  Once one migration is finished the other processes would be able to query `phinxlog` and see that all migrations are already performed.

Obtaining a read lock on a table is released when a transaction is started (in MySQL), so the transaction in `executeMigration()` would have to be transferred "up" a level to `migrate()`.

Whether the example scenario is considered best practices or not, the migration tool should put safety of the production database above all else and handle accidental concurrent migration attempts in a consistent and predictable manner (i.e. obtaining a lock that queues reads from other connections so only 1 migration can proceed at a time)

```
    public function migrate($environment, $version = null, $fake = false)
    {
        $migrations = $this->getMigrations($environment);
        $env = $this->getEnvironment($environment);

        if (!$fake) {
            // begin the transaction if the adapter supports it
            if ($env->getAdapter()->hasTransactions()) {
                $env->getAdapter()->beginTransaction();
            }
         }
        //obtain exclusive lock on phinxlog to serialize concurrent migrations
       
        $versions = $env->getVersions();
        $current = $env->getCurrentVersion();
        ...
```

**Comment by dereuromark** [comment]
PR welcome

**Comment by othercorey** [comment]
Closing due to lack of activity or out of scope. If there is new interest, please open a PR.
